kiwmi/desktop/view.c

```diff
@@ -259,3 +284,191 @@ view_create(
+}
+
+static void
+view_child_surface_destroy_notify(
+    struct wl_listener *listener,
+    void *UNUSED(data))
+{
+    struct kiwmi_view_child *child =
+        wl_container_of(listener, child, surface_destroy);
+    view_child_destroy(child);
+}
+
+static void
+view_child_commit_notify(struct wl_listener *listener, void *UNUSED(data))
+{
+    struct kiwmi_view_child *child = wl_container_of(listener, child, commit);
+    view_child_damage(child);
+}
+
+static void
+view_child_new_subsurface_notify(struct wl_listener *listener, void *data)
+{
+    struct kiwmi_view_child *child =
+        wl_container_of(listener, child, new_subsurface);
+    struct wlr_subsurface *subsurface = data;
+    view_child_subsurface_create(child, child->view, subsurface);
+}
+
+static void
+view_child_map_notify(struct wl_listener *listener, void *UNUSED(data))
+{
+    struct kiwmi_view_child *child = wl_container_of(listener, child, map);
+    child->mapped                  = true;
+    view_child_damage(child);
+}
+
+static void
+view_child_unmap_notify(struct wl_listener *listener, void *UNUSED(data))
+{
+    struct kiwmi_view_child *child = wl_container_of(listener, child, unmap);
+    child->mapped                  = false;
+    view_child_damage(child);
+}
+
+static void
+view_child_view_unmap_notify(struct wl_listener *listener, void *UNUSED(data))
+{
+    struct kiwmi_view_child *child =
+        wl_container_of(listener, child, view_unmap);
+    child->mapped = false;
+    view_child_damage(child);
+}
+
+struct kiwmi_view_child *
+view_child_create(
+    struct kiwmi_view *view,
+    struct wlr_surface *wlr_surface,
+    enum kiwmi_view_child_type type,
+    const struct kiwmi_view_child_impl *impl)
+{
+    struct kiwmi_view_child *child = calloc(1, sizeof(*child));
+    if (!child) {
+        wlr_log(WLR_ERROR, "Failed to allocate view_child");
+        return NULL;
+    }
+
+    child->type        = type;
+    child->impl        = impl;
+    child->view        = view;
+    child->wlr_surface = wlr_surface;
+
+    wl_list_init(&child->link);
+    wl_list_init(&child->children);
+    child->mapped = false;
+
+    child->commit.notify = view_child_commit_notify;
+    wl_signal_add(&wlr_surface->events.commit, &child->commit);
+
+    child->surface_destroy.notify = view_child_surface_destroy_notify;
+    wl_signal_add(&wlr_surface->events.destroy, &child->surface_destroy);
+
+    child->new_subsurface.notify = view_child_new_subsurface_notify;
+    wl_signal_add(&wlr_surface->events.new_subsurface, &child->new_subsurface);
+
+    child->view_unmap.notify = view_child_view_unmap_notify;
+    wl_signal_add(&view->events.unmap, &child->view_unmap);
+
+    // Let the caller add the listeners to the respective events
+    child->map.notify = view_child_map_notify;
+    wl_list_init(&child->map.link);
+
+    child->unmap.notify = view_child_unmap_notify;
+    wl_list_init(&child->unmap.link);
+
+    wl_list_init(&child->new_popup.link);
```

Comment:
`new_popup` is only used in `xdg_popup`, so I’d say that at least it should (though there’s also the `calloc` 🤷)

I’d rather keep it for `owner_destroy` as well, since a new `view_child` could potentially not have it.

Reply:
Ah, I see. Yeah with the remove stuff better keep it. Good catch!